### PR TITLE
Adding TimeProvider CreateCancellationTokenSource extension method

### LINKED ISSUE #85080: [API Proposal]: Introduce TimeProvider extension method to create a CTS to be cancelled after a specified time period.
### Background and motivation

A new constructor called `CancellationTokenSource(TimeSpan, TimeProvider)` has been introduced for creating a CTS object that automatically gets cancelled after a specified period using the TimeProvider. However, this constructor is only available on .NET 8.0. For users who are using TimeProvider on down-level versions, they will need to write their own code to achieve the same functionality. An example of such code is given below.

```C#
        private static void CancelAfter(TimeProvider provider, CancellationTokenSource cts, TimeSpan delay)
        {
            if (provider == TimeProvider.System)
            {
                cts.CancelAfter(delay);
            }
            else
            {
                ITimer timer = provider.CreateTimer(s => ((CancellationTokenSource)s).Cancel(), cts, delay, Timeout.InfiniteTimeSpan);
                cts.Token.Register(t => ((ITimer)t).Dispose(), timer);
            }
        }
```

The experience would be unsatisfactory when users want to write a ns2.0 library code that works in all supported framework versions.

### API Proposal

```csharp
namespace System.Threading.Tasks
{
    public static class TimeProviderTaskExtensions
    {

        /// <summary>Initializes a new instance of the <see cref="CancellationTokenSource"/> class that will be canceled after the specified <see cref="TimeSpan"/>. </summary>
        /// <param name="timeProvider">The <see cref="TimeProvider"/> with which to interpret the <paramref name="delay"/>. </param>
        /// <param name="delay">The time interval to wait before canceling this <see cref="CancellationTokenSource"/>. </param>
        /// <exception cref="ArgumentOutOfRangeException"><paramref name="delay"/>'s <see cref="TimeSpan.TotalMilliseconds"/> is less than -1 or greater than <see cref="uint.MaxValue"/> - 1. </exception>
        /// <remarks>
        /// The countdown for the delay starts during the call to the constructor.  When the delay expires,
        /// the constructed <see cref="CancellationTokenSource"/> is canceled if it has
        /// not been canceled already. Subsequent calls to CancelAfter will reset the delay for the constructed
        /// <see cref="CancellationTokenSource"/> if it has not been canceled already.
        /// </remarks>
        public static CancellationTokenSource CreateCancellationTokenSource(this TimeProvider timeProvider, TimeSpan delay) ;
    }
}
```

### API Usage

```csharp

CancellationTokenSource  cts = TimeProvider.System.CreateCancellationTokenSource(TimeSpan.FromMilliseconds(100));

....

cts.Task.WaitAsync(CancellationToken.None);
```

### Alternative Designs

_No response_

### Risks

_No response_

### PR DESCRIPTION
**Pull request by tarekgh**
Fixes https://github.com/dotnet/runtime/issues/85080

### CODE REVIEW

**Comment by tarekgh** [comment]
CC @sebastienros

**Review by sebastienros — APPROVED** [review]

File: src/libraries/Microsoft.Bcl.TimeProvider/src/System/Threading/Tasks/TimeProviderTaskExtensions.cs [hunk 1]
```diff
@@ -218,5 +218,46 @@ public static async Task<TResult> WaitAsync<TResult>(this Task<TResult> task, Ti
...
+        /// <param name="timeProvider">The <see cref="TimeProvider"/> with which to interpret the <paramref name="delay"/>. </param>
+        /// <param name="delay">The time interval to wait before canceling this <see cref="CancellationTokenSource"/>. </param>
+        /// <exception cref="ArgumentOutOfRangeException"> The <paramref name="delay"/> is negative and not equal to <see cref="Timeout.InfiniteTimeSpan" /> or greater than maximum allowed timer duration.</exception>
+        /// <returns><see cref="CancellationTokenSource"/> that will be canceled after the specified <paramref name="delay"/>.</returns>
+        /// <remarks>
+        /// The countdown for the delay starts during the call to the constructor. When the delay expires,
+        /// the constructed <see cref="CancellationTokenSource"/> is canceled if it has
+        /// not been canceled already.
+        /// If running on framework version prior to .NET 8.0, there is a constraint when invoking <see cref="CancellationTokenSource.CancelAfter(TimeSpan)"/> on the resultant object.
```
Comment (danmoseley) [inline src/libraries/Microsoft.Bcl.TimeProvider/src/System/Threading/Tasks/TimeProviderTaskExtensions.cs:231]:
```suggestion
        /// If running on .NET versions earlier than .NET 8.0, there is a constraint when invoking <see cref="CancellationTokenSource.CancelAfter(TimeSpan)"/> on the resultant object.
```

File: src/libraries/Microsoft.Bcl.TimeProvider/src/System/Threading/Tasks/TimeProviderTaskExtensions.cs [hunk 2]
```diff
@@ -218,5 +218,46 @@ public static async Task<TResult> WaitAsync<TResult>(this Task<TResult> task, Ti
...
+
+            var cts = new CancellationTokenSource();
+
+            if (timeProvider == TimeProvider.System)
+            {
+                cts.CancelAfter(delay);
+                return cts;
+            }
+
+            ITimer timer = timeProvider.CreateTimer(s => ((CancellationTokenSource)s).Cancel(), cts, delay, Timeout.InfiniteTimeSpan);
```
Comment (stephentoub) [inline src/libraries/Microsoft.Bcl.TimeProvider/src/System/Threading/Tasks/TimeProviderTaskExtensions.cs:257]:
```suggestion
            ITimer timer = timeProvider.CreateTimer(s =>
            {
                try
                {
                    ((CancellationTokenSource)s).Cancel();
                }
                catch (ObjectDisposedException) { }
            }, cts, delay, Timeout.InfiniteTimeSpan);
```
Please also add a test that:
1. Creates the CTS with a delay that we control via the time provider.
2. Disposes the CTS.
3. Then fires the timer.  We want to validate the process doesn't crash.

File: src/libraries/Microsoft.Bcl.TimeProvider/src/System/Threading/Tasks/TimeProviderTaskExtensions.cs (hunk 1 above)
Comment (stephentoub) [inline src/libraries/Microsoft.Bcl.TimeProvider/src/System/Threading/Tasks/TimeProviderTaskExtensions.cs:231]: If we don't already have such statements elsewhere, we should also include in the docs that these extensions should only be used when building against pre-.NET 8 surface area.  For code building against .NET 8+, they shouldn't be using this package at all.

Comment (JamesNK) [inline src/libraries/Microsoft.Bcl.TimeProvider/src/System/Threading/Tasks/TimeProviderTaskExtensions.cs:231]: XML doc comments don't care about whitespace. You need to wrap paragraphs in `<para></para>` tags if you want the content to appear on a new line.

For example: https://github.com/dotnet/aspnetcore/blob/fca80883d95102eeb96e6fe9a68b170d54a4614c/src/Http/Http.Features/src/IFormFeature.cs#L16-L23

Comment (eiriktsarpalis) [reply src/libraries/Microsoft.Bcl.TimeProvider/src/System/Threading/Tasks/TimeProviderTaskExtensions.cs:231]: We use the [porting tool](https://github.com/dotnet/api-docs-sync) to port documentation from source to the dotnet-api-docs repo. I believe remarks sections do have some degree of whitespace significance there so `<para />` tags aren't all that common in dotnet/runtime source code.

cc @carlossanlop

**Pull request merged**